man/shutdown.8
'\" -*- coding: UTF-8 -*-
.\" Copyright (C) 1998-2003 Miquel van Smoorenburg.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
.\"
.\"{{{}}}
.\"{{{  Title
.TH SHUTDOWN 8 "November 12, 2003" "sysvinit " "Linux System Administrator's Manual"
.\"}}}
.\"{{{  Name
.SH NAME
shutdown \- bring the system down
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B /sbin/shutdown
.RB [ \-akrhPHfFncqQ ]
.RB [ \-t
.IR sec ]
.I time
.RI [ "warning message" ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
\fBshutdown\fP brings the system down in a secure way.  All logged-in users are
notified that the system is going down, and \fBlogin\fP(1) is blocked.
It is possible to shut the system down immediately or after a specified delay.
All processes are first notified that the system is going down by the
signal \fBSIGTERM\fP.  This gives programs like \fBvi\fP(1)
the time to save the file being edited,
mail and news processing programs a chance to exit cleanly, etc.
\fBshutdown\fP does its job by signalling the \fBinit\fP process,
asking it to change the runlevel.
Runlevel \fB0\fP is used to halt the system, runlevel \fB6\fP is used
to reboot the system, and runlevel \fB1\fP is used to put to system into
a state where administrative tasks can be performed; this is the default
if neither the \fB\-h\fP or \fB\-r\fP flag is given to \fBshutdown\fP.
To see which actions are taken on halt or reboot see the appropriate
entries for these runlevels in the file \fI/etc/inittab\fP.
.\"}}}
.\"{{{  Options
.SH OPTIONS
.\"{{{  -a
.IP "\fB\-a\fP
Use \fI/etc/shutdown.allow\fP.
.\"}}}
.\"{{{  -k
.IP \fB\-k\fP
Don't really shutdown; only send the warning messages to everybody.
.\"}}}
.\"{{{  -r
.IP \fB\-r\fP
Reboot after shutdown.
.\"}}}
.\"{{{  -h
.IP \fB\-h\fP
Halt or power off after shutdown.
Usually used with the \fB\-P\fP or \fB\-H\fP flags,
depending on whether we want to poweroff or
simply stop the operating system.
.\"}}}
.\"{{{  -P
.IP \fB\-P\fP
Modifier to the \fB\-h\fP flag. Halt action is to turn off the power.
Must be used with the \fB\-h\fP flag.
.\"}}}
.\"{{{  -H
.IP \fB\-H\fP
Modifier to the \fB\-h\fP flag.  Halt action is to halt or drop into boot
monitor on systems that support it.  Must be used with the \fB\-h\fP flag.
Halting is often used to run through the shutdown process and leave
output on the screen for debugging purposes. Or when the user wants the OS to
stop, but leave the power on. To power off at the end of the shutdown sequence
use the \fB\-P\fP modifier instead.
.\"}}}
.\"{{{  -f
.IP \fB\-f\fP
Skip \fBfsck\fP(8) on reboot.
.\"}}}
.\"{{{  -F
.IP \fB\-F\fP
Force \fBfsck\fP(8) on reboot.
.\"}}}
.\"{{{  -n
.IP \fB\-n\fP
[DEPRECATED] Don't call \fBinit\fP(8) to do the shutdown but do it ourself.
The use of this option is discouraged, and its results are not always what
you'd expect.
.\"}}}
.\"{{{  -c
.IP \fB\-c\fP
Cancel a waiting shutdown. (\fBshutdown now\fP is no longer waiting.) With
this option it is of course not possible to give the time argument, but
you can enter explanatory message arguments on the command line that
will be sent to all users.
.\"{{{  -q
.IP \fB\-q
Reduce the number of warnings \fBshutdown\fP displays.
Usually \fBshutdown\fP displays
warnings every 15 minutes and then every minute in the last 10 minutes of the
countdown until \fItime\fP is reached. When \fB\-q\fP is specified
\fBshutdown\fP only warns at 60 minute intervals, at the 10 minute mark,
at the 5 minute mark, and when the shutdown process actually happens.
.\"{{{  -Q
.IP \fB\-Q
Silence warnings prior to shutting down. Usually \fBshutdown\fP displays
warnings every 15 minutes and then every minute in the last 10 minutes of the
countdown until \fItime\fP is reached. When \fB\-Q\fP is specified
\fBshutdown\fP only warns when the shutdown process actually happens. All
other warning intervals are suppressed.
.\"}}}
.\"{{{  -t sec
.IP "\fB\-t\fP \fIsec\fP"
Tell \fBinit\fP(8) to wait \fIsec\fP seconds between sending all processes the
warning (\fBSIGTERM\fP) and the kill signal (\fBSIGKILL\fP),
before changing to another runlevel.
The default time, if no value is specified, between these two signals is
three seconds.
Warning: when \fBshutdown\fP calls \fBinit\fP(8) to perform the shutdown
(the default behavior),
\fBinit\fP(8) checks to see if all processes have terminated
and will stop waiting early once its children have all terminated.
When \fBshutdown\fP is called with the \fB\-n\fP flag,
it waits the full time specified
(or three seconds) even if all other processes have terminated.
.\"}}}
.\"{{{  time
.IP \fItime\fP
When to shutdown.
.\"}}}
.\"{{{  warning-message
.IP "\fIwarning message\fP"
Message to send to all users.
.\"}}}
.PP
The \fItime\fP argument can have different formats.  First, it can be an
absolute time in the format \fIhh:mm\fP, in which \fIhh\fP is the hour
(1 or 2 digits) and \fImm\fP is the minute of the hour (in two digits).
Second, it can be in the format \fB+\fP\fIm\fP, in which \fIm\fP is the
number of minutes to wait.  Third, it can be in the format \fB+\fP\fIhh:mm\fP,
in which \fIhh:mm\fP is the number of hours and minutes to wait.
The word \fBnow\fP is an alias for \fB+0\fP.
.PP
If \fBshutdown\fP is called with a delay, it will create the advisory file
.I /etc/nologin
which causes programs such as \fBlogin\fP(1) to not allow new user
logins. This file is created five minutes before the shutdown sequence
starts. \fBshutdown\fP removes this file if it is stopped before it
can signal init (i.e., it is cancelled or something goes wrong).
It also removes it before calling \fBinit\fP(8) to change the runlevel.
.PP
The \fB\-f\fP flag means `reboot fast'.  This only creates an advisory
file \fI/fastboot\fP which can be tested by the system when it comes
up again.  The boot rc file can test if this file is present, and decide not
to run \fBfsck\fP(8) since the system has been shut down in the proper way.
After that, the boot process should remove \fI/fastboot\fP.
.PP
The \fB\-F\fP flag means `force fsck'.  This only creates an advisory
file \fI/forcefsck\fP which can be tested by the system when it comes
up again.  The boot rc file can test if this file is present, and decide

doc/Changelog
sysvinit (3.10) released; urgency=low

    *  When the user executes "machinectl stop", systemd sends SIGRTMIN+4 to PID 1 
       in the container, and expects that to initiate a graceful shutdown (power-off).
       SysV init now catches this signal and initiates a shutdown (shutdown -hP now).
       - floppym provided patch to accomplish this.

    * Fix issue in bootlogd which could cause the service to enter an endless loop
      (and use too much CPU) when it is able to open a device for writing, but not actually
      able to write to it. This resulted in bootlogd closing and re-opening the device over
      and over. Now bootlogd should simply fail gracefully when it cannot write to an open
      file/device.

    * Fix formatting in shutdown.8 manual page. Cleaned up whitespace and special characters.


sysvinit (3.09) released; urgency=low

    * Patch for man/Makefile to fix the clean recipe.
      Provided by Lucas Nussabaum and Mark Hindley

    * On Linux systems, allow reboot command to pass a message
      to the system firmware during the restart. This is
      accomplished with the -m flag.

    * Patch from kraj which allows hddown to compile
      when musl is the C library.

sysvinit (3.08) released; urgency=low
    * floppym provided patch which causes the halt command
      to call "shutdown -h -H" instead of "shutdown -h" when
      halt is invoked without parameters. This forces the shutdown
      command to set the INIT_HALT variable and assume, unless other
      conditions apply, that the "halt" call really wants to halt the
      machine and INIT_HALT should be set. In other words we
      assume halt wants to halt unless told otherwise.
      Addresses downstream Gentoo bug ID 911257.

    * Applied a patch from floppm which adds kexec option to the halt command.

    * Updated halt documentation and help output to display parameters in
      alphabetical order.


sysvinit (3.07) released; urgency=low
    * Fixed killall5 so that processes in the omit list are
      not sent any signals, including SIGSTOP.
    * Fixed usage message for killall5 to be more accurate.
    * pidof was not returning PIDs of programs which were launched
      using a symbolic link. (ie /tmp/sleep when /tmp/sleep links to /usr/bin/sleep).
      This is now fixed as we check both the realpath and symbolic path for processes.
      In other words, "pidof /tmp/sleep" and "pidof /usr/bin/sleep" will return
      the same PIDs when /tmp/sleep is a symbolic link to /usr/bin/sleep.
    * Fixed memory initialization error in pidof. Fix provided by Markus Fischer.
    * Accepted patch from Mark Hindley  which avoids clearing realpath information
      in pidof when trying to find matching executables.


sysvinit (3.06) released; urgency=low
    * Mark Hindley fixed typo in es.po
    * Mark Hindley cleaned up translation code in src/Makefile. 
    * Drop sulogin from Debian build. Removed libcrypt-dev dependency.
    * Fixed pt translation pages which were failing due to mis-matched
      open/close tags.
    * Makefile now respects ROOT prefix when setting up pidof-to-killall5
      symbolic link.
    * Removed redundant translation files from man directory.
    * Makefile now respects DESTDIR. User can specify either ROOT= or DESTDIR=
      to set install prefix.


sysvinit (3.05) released; urgency=low
    * Helge Kreutzmann provided updated Makefile for

src/shutdown.c
/*
 * shutdown.c	Shut the system down.
 *
 * Usage:	shutdown [-krhfnc] time [warning message]
 *		  -k: don't really shutdown, only warn.
 *		  -r: reboot after shutdown.
 *		  -h: halt after shutdown.
 *		  -f: do a 'fast' reboot (skip fsck).
 *		  -F: Force fsck on reboot.
 *		  -n: do not go through init but do it ourselves.
 *		  -c: cancel an already running shutdown.
 *		  -t secs: delay between SIGTERM and SIGKILL for init.
 *
 * Author:	Miquel van Smoorenburg, miquels@cistron.nl
 *
 * Version:	@(#)shutdown  2.86-1  31-Jul-2004  miquels@cistron.nl
 *
 *		This file is part of the sysvinit suite,
 *		Copyright (C) 1991-2004 Miquel van Smoorenburg.
 *
 *		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 2 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License
 *		along with this program; if not, write to the Free Software
 *		Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */
#ifndef _GNU_SOURCE
# define _GNU_SOURCE	/* otherwise `extern char **environ' is missed */
#endif
#ifndef ACCTON_OFF
# define ACCTON_OFF	0
#endif
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#ifdef __linux__
#include <sys/sysmacros.h>   /* brought in my LFS patch */
#endif
#include <time.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>
#include <stdio.h> 
#include <signal.h>
#include <fcntl.h>
#include <stdarg.h>
#ifdef __FreeBSD__
#include <utmpx.h>
#else
#include <utmp.h>
#endif
#include <syslog.h>
#include "paths.h"
#include "reboot.h"
#include "initreq.h"
#include "init.h"

#ifdef __FreeBSD__
extern char **environ;
#endif

#define MESSAGELEN	256
#define STATELEN        64
#define WHEN_SIZE       64

/* Whether we should warn system is shutting down */
#define QUIET_FULL 2
#define QUIET_PARTIAL 1
#define QUIET_NONE 0

int dontshut = 0;	/* Don't shutdown, only warn	*/
char down_level[2];	/* What runlevel to go to.	*/
int dosync = 1;		/* Sync before reboot or halt	*/
int fastboot = 0;	/* Do a 'fast' reboot		*/
int forcefsck = 0;	/* Force fsck on reboot		*/
char message[MESSAGELEN];	/* Warning message	*/
char *sltime = 0;	/* Sleep time			*/
char newstate[STATELEN];	/* What are we gonna do		*/
int doself = 0;		/* Don't use init		*/
int got_alrm = 0;

char *clean_env[] = {
	"HOME=/",
	"PATH=" PATH_DEFAULT,
	"TERM=dumb",
	"SHELL=/bin/sh",
	NULL,
};

/* From "utmp.c" */
extern void write_wtmp(char *user, char *id, int pid, int type, char *line);

/*
 *	Sleep without being interrupted.
 */
void hardsleep(int secs)
{
	struct timespec ts, rem;

	ts.tv_sec = secs;
	ts.tv_nsec = 0;

	while(nanosleep(&ts, &rem) < 0 && errno == EINTR)
		ts = rem;
}

/*
 *	Break off an already running shutdown.
 */
# ifdef __GNUC__
void stopit(int sig __attribute__((unused)))
# else
void stopit(int sig)
# endif

{
	unlink(NOLOGIN);
	unlink(FASTBOOT);
	unlink(FORCEFSCK);
	unlink(SDPID);
	printf("\r\nShutdown cancelled.\r\n");
	exit(0);
}

/*
 *	Show usage message.
 */
void usage(void)
{
	fprintf(stderr,
	"Usage:\t  shutdown [-akrhPHfFnc] [-t sec] time [warning message]\n"
	"\t\t  -a:      use /etc/shutdown.allow\n"
	"\t\t  -k:      don't really shutdown, only warn.\n"
	"\t\t  -r:      reboot after shutdown.\n"
	"\t\t  -h:      halt after shutdown.\n"
	"\t\t  -P:      halt action is to turn off power.\n"
        "\t\t           can only be used along with -h flag.\n"
	"\t\t  -H:      halt action is to just halt.\n"
        "\t\t           can only be used along with -h flag.\n"
	"\t\t  -f:      do a 'fast' reboot (skip fsck).\n"
	"\t\t  -F:      Force fsck on reboot.\n"
	"\t\t  -n:      do not go through \"init\" but go down real fast.\n"
	"\t\t  -c:      cancel a running shutdown.\n"
        "\t\t  -q:      quiet mode - display fewer shutdown warnings.\n"
        "\t\t  -Q:      full quiet mode - display only final shutdown warning.\n"
	"\t\t  -t secs: delay between warning and kill signal.\n"
	"\t\t  ** the \"time\" argument is mandatory! (try \"now\") **\n");
	exit(1);
}


void alrm_handler(int sig)
{
	got_alrm = sig;
}


/*
 *	Set environment variables in the init process.
 */
int init_setenv(char *name, char *value)
{
	struct init_request	request;
	struct sigaction	sa;
	int			fd;
	size_t			nl, vl;

	memset(&request, 0, sizeof(request));
	request.magic = INIT_MAGIC;
	request.cmd = INIT_CMD_SETENV;
	nl = strlen(name);
	vl = value ? strlen(value) : 0;

	if (nl + vl + 3 >= (int)sizeof(request.i.data))
		return -1;

	memcpy(request.i.data, name, nl);
	if (value) {
		request.i.data[nl] = '=';
		memcpy(request.i.data + nl + 1, value, vl);
	}

        /*
	 *	Open the fifo and write the command.
         *	Make sure we don't hang on opening /run/initctl
	 */
	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = alrm_handler;
	sigaction(SIGALRM, &sa, NULL);
	got_alrm = 0;
	alarm(3);
	if ((fd = open(INIT_FIFO, O_WRONLY)) >= 0) {
		ssize_t p = 0;
		size_t s  = sizeof(request);
		void *ptr = &request;
		while (s > 0) {
			p = write(fd, ptr, s);
			if (p < 0) {
				if (errno == EINTR || errno == EAGAIN)
					continue;
				break;
			}
			ptr += p;
			s -= p;
		}
		close(fd);
		alarm(0);
		return 0;
	}
                                                                                
	fprintf(stderr, "shutdown: ");
	if (got_alrm) {
		fprintf(stderr, "timeout opening/writing control channel %s\n",
			INIT_FIFO);
	} else {
		perror(INIT_FIFO);
	}
	return -1;
}


/*
 *	Tell everyone the system is going down in 'mins' minutes.
 */
void issue_warn(int mins)
{
	char buf[MESSAGELEN + sizeof(newstate) + 1];
	int len;

	buf[0] = 0;
	strncpy(buf, message, MESSAGELEN);
	len = strlen(buf);

	if (mins == 0)
		snprintf(buf + len, sizeof(buf) - len,
			"\rThe system is going down %s NOW!\r\n",
			newstate);
	else
  		snprintf(buf + len, sizeof(buf) - len,
			"\rThe system is going DOWN %s in %d minute%s!\r\n",
				newstate, mins, mins == 1 ? "" : "s");
	wall(buf, 0);
}

/*
 *	Create the /etc/nologin file.
 */
void donologin(int min)
{
	FILE *fp;
	time_t t;

	time(&t);
	t += 60 * min;

	if ((fp = fopen(NOLOGIN, "w")) != NULL) {
  		fprintf(fp, "\rThe system is going down on %s\r\n", ctime(&t));
  		if (message[0]) fputs(message, fp);
  		fclose(fp);
	}
}

/*
 *	Spawn an external program.
 */
int spawn(int noerr, char *prog, ...)
{
	va_list	ap;
	pid_t	pid, rc;
	int	i;
	char	*argv[8];

	i = 0;
	while ((pid = fork()) < 0 && i < 10) {
		perror("fork");
		sleep(5);
		i++;
	}

	if (pid < 0) return -1;

	if (pid > 0) {
		while((rc = wait(&i)) != pid)
			if (rc < 0 && errno == ECHILD)
				break;
		return (rc == pid) ? WEXITSTATUS(i) : -1;
	}

	if (noerr) fclose(stderr);

	argv[0] = prog;
	va_start(ap, prog);
	for (i = 1; i < 7 && (argv[i] = va_arg(ap, char *)) != NULL; i++)
		;
	argv[i] = NULL;
	va_end(ap);

	if (chdir("/"))
		exit(1);
	environ = clean_env;

	execvp(argv[0], argv);
	perror(argv[0]);
	exit(1);

	/*NOTREACHED*/
	return 0;
}

/*
 *	Kill all processes, call /etc/init.d/halt (if present)
 */
void fastdown()
{
	int do_halt = (down_level[0] == '0');
	int i;
#if 0
	char cmd[128];
	char *script;

	/*
	 *	Currently, the halt script is either init.d/halt OR rc.d/rc.0,
	 *	likewise for the reboot script. Test for the presence
	 *	of either.
	 */
	if (do_halt) {
		if (access(HALTSCRIPT1, X_OK) == 0)
			script = HALTSCRIPT1;
		else
			script = HALTSCRIPT2;
	} else {
		if (access(REBOOTSCRIPT1, X_OK) == 0)
			script = REBOOTSCRIPT1;
		else
			script = REBOOTSCRIPT2;
	}
#endif

	/* First close all files. */
	for(i = 0; i < 3; i++)
		if (!isatty(i)) {
			close(i);
			open("/dev/null", O_RDWR);
		}
	for(i = 3; i < 20; i++) close(i);
	close(255);

	/* First idle init. */
	if (kill(1, SIGTSTP) < 0) {
		fprintf(stderr, "shutdown: can't idle init: %s.\r\n", strerror(errno));
		exit(1);
	}

	/* Kill all processes. */
	fprintf(stderr, "shutdown: sending all processes the TERM signal...\r\n");
	kill(-1, SIGTERM);
	sleep(sltime ? atoi(sltime) : WAIT_BETWEEN_SIGNALS);
	fprintf(stderr, "shutdown: sending all processes the KILL signal.\r\n");
	(void) kill(-1, SIGKILL);

#if 0
	/* See if we can run /etc/init.d/halt */
	if (access(script, X_OK) == 0) {
		spawn(1, cmd, "fast", NULL);
		fprintf(stderr, "shutdown: %s returned - falling back "
				"on default routines\r\n", script);
	}
#endif

	/* script failed or not present: do it ourself. */
	/* Give init the chance to collect zombies. */
        /* sleep(1); */

	/* Record the fact that we're going down */
	write_wtmp("shutdown", "~~", 0, RUN_LVL, "~~");

	/* This is for those who have quota installed. */
#if defined(ACCTON_OFF)
# if (ACCTON_OFF > 1) && (_BSD_SOURCE || (_XOPEN_SOURCE && _XOPEN_SOURCE < 500))
	/* This is an alternative way to disable accounting, saving a fork() */
	if (acct(NULL))
		fprintf(stderr, "shutdown: can not stop process accounting: %s.\r\n", strerror(errno));
# elif (ACCTON_OFF > 0)
	spawn(1, "accton", "off", NULL);
# else
	spawn(1, "accton", NULL);
# endif
#endif
	spawn(1, "quotaoff", "-a", NULL);

	sync();
	fprintf(stderr, "shutdown: turning off swap\r\n");
	spawn(0, "swapoff", "-a", NULL);
	fprintf(stderr, "shutdown: unmounting all file systems\r\n");
	spawn(0, "umount", "-a", NULL);

	/* We're done, halt or reboot now. */
	if (do_halt) {
		fprintf(stderr, "The system is halted. Press CTRL-ALT-DEL "
				"or turn off power\r\n");
		init_reboot(BMAGIC_HALT);
		exit(0);
	}

	fprintf(stderr, "Please stand by while rebooting the system.\r\n");
	init_reboot(BMAGIC_REBOOT);
	exit(0);
}

/*
 *	Go to runlevel 0, 1 or 6.
 */
void issue_shutdown(char *halttype)
{
	char	*args[8];
	int	argp = 0;
	int	do_halt = (down_level[0] == '0');

	/* Warn for the last time */
	issue_warn(0);
	if (dontshut) {
		hardsleep(1);
		stopit(0);
	}
	openlog("shutdown", LOG_PID, LOG_USER);
	if (do_halt)
  		syslog(LOG_NOTICE, "shutting down for system halt");
	else
		syslog(LOG_NOTICE, "shutting down for system reboot");
	closelog();

	/* See if we have to do it ourself. */
	if (doself) fastdown();

	/* Create the arguments for init. */
	args[argp++] = INIT;
	if (sltime) {
		args[argp++] = "-t";
		args[argp++] = sltime;
	}
	args[argp++] = down_level;
	args[argp]   = (char *)NULL;

	unlink(SDPID);
	unlink(NOLOGIN);

	/* Now execute init to change runlevel. */
	sync();
	init_setenv("INIT_HALT", halttype);
	execv(INIT, args);

	/* Oops - failed. */
	fprintf(stderr, "\rshutdown: cannot execute %s\r\n", INIT);
	unlink(FASTBOOT);
	unlink(FORCEFSCK);
	init_setenv("INIT_HALT", NULL);
	openlog("shutdown", LOG_PID, LOG_USER);
	syslog(LOG_NOTICE, "shutdown failed");
	closelog();
	exit(1);
}

/*
 *	returns if a warning is to be sent for wt
 */
static int needwarning(int wt, int quiet_mode)
{
	int ret;

        if (quiet_mode == QUIET_FULL) return FALSE;
        else if (quiet_mode == QUIET_PARTIAL)
        {
            if (wt == 10)
               return TRUE;
            else if (wt == 5)
               return TRUE;
            else if ( (wt % 60) == 0)
               return TRUE;
            else
               return FALSE;
        }
        /* no silence setting, print lots of warnings */
	if (wt < 10)
		ret = 1;
	else if (wt < 60)
		ret = (wt % 15 == 0);
	else if (wt < 180)
		ret = (wt % 30 == 0);
	else
		ret = (wt % 60 == 0);

	return ret;
}

/*
 *	Main program.
 *	Process the options and do the final countdown.
 */
int main(int argc, char **argv)
{
	FILE			*fp;
	extern int		getopt();
	extern int		optind; 
	struct sigaction	sa;
	struct tm		*lt;
	struct stat		st;
	struct utmp		*ut;
	time_t			t, target_time;
	char			*halttype;
	char			*downusers[32];
	char			buf[128];
	char			term[UT_LINESIZE + 6];
	char			*sp;
	char			when[WHEN_SIZE];
	int			c, i, wt;
	int			hours, mins;
	int			didnolog = 0;
	int			cancel = 0;
	int			useacl = 0;
	int			pid = 0;
	int			user_ok = 0;
        int quiet_level = QUIET_NONE;   /* Whether to display shutdown warning */

	/* We can be installed setuid root (executable for a special group) */
	/* 
        This way is risky, do error check on setuid call.
        setuid(geteuid());
        */
        errno = 0;
        if (setuid(geteuid()) == -1) {
            fprintf(stderr, "%s (%d): %s\n", __FILE__, __LINE__, strerror(errno));
            abort();
	}

	if (getuid() != 0) {
  		fprintf(stderr, "shutdown: you must be root to do that!\n");
		usage();
  		exit(1);
	}
	strcpy(down_level, "1");
	halttype = NULL;
        memset(when, '\0', WHEN_SIZE);

	/* Process the options. */
	while((c = getopt(argc, argv, "HPacqQkrhnfFyt:g:i:")) != EOF) {
  		switch(c) {
			case 'H':
				halttype = "HALT";
				break;
			case 'P':
				halttype = "POWEROFF";
				break;
			case 'a': /* Access control. */
				useacl = 1;
				break;
			case 'c': /* Cancel an already running shutdown. */
				cancel = 1;
				break;
  			case 'k': /* Don't really shutdown, only warn.*/
  				dontshut = 1;
  				break;
  			case 'r': /* Automatic reboot */
				down_level[0] = '6';
  				break;
  			case 'h': /* Halt after shutdown */
				down_level[0] = '0';
  				break;
  			case 'f': /* Don't perform fsck after next boot */
  				fastboot = 1;
  				break;
  			case 'F': /* Force fsck after next boot */
  				forcefsck = 1;
  				break;
			case 'n': /* Don't switch runlevels. */
				doself = 1;
				break;
			case 't': /* Delay between TERM and KILL */
				sltime = optarg;
				break;
                        case 'q': /* put into somewhat quiet mode */
                                quiet_level = QUIET_PARTIAL;
                                break;
                        case 'Q': /* put into full quiet mode */
                                quiet_level = QUIET_FULL;
                                break;
			case 'y': /* Ignored for sysV compatibility */
				break;
			case 'g': /* sysv style to specify time. */
				strncpy(when, optarg, WHEN_SIZE - 1);
				break;
			case 'i': /* Level to go to. */
				if (!strchr("0156aAbBcCsS", optarg[0])) {
					fprintf(stderr,
					"shutdown: `%s': bad runlevel\n",
					optarg);
					exit(1);
				}
				down_level[0] = optarg[0];
				break;
  			default:
  				usage();
  				break;	
  		}
	}

	if (NULL != halttype && down_level[0] != '0') {
		fprintf(stderr, "shutdown: -H and -P flags can only be used along with -h flag.\n");
		usage();
  		exit(1);
	}

	/* Do we need to use the shutdown.allow file ? */
	if (useacl && (fp = fopen(SDALLOW, "r")) != NULL) {

		/* Read /etc/shutdown.allow. */
		i = 0;
		while(fgets(buf, 128, fp)) {
			if (buf[0] == '#' || buf[0] == '\n') continue;
			if (i > 31) continue;
			for(sp = buf; *sp; sp++) if (*sp == '\n') *sp = 0;
			downusers[i++] = strdup(buf);
		}
		if (i < 32) downusers[i] = 0;
		fclose(fp);

		/* Now walk through /var/run/utmp to find logged in users. */
		while(!user_ok && (ut = getutent()) != NULL) {

			/* See if this is a user process on a VC. */
			if (ut->ut_type != USER_PROCESS) continue;
			sprintf(term, "/dev/%.*s", UT_LINESIZE, ut->ut_line);
			if (stat(term, &st) < 0) continue;
#ifdef major /* glibc */
			if (major(st.st_rdev) != 4 ||
			    minor(st.st_rdev) > 63) continue;
#else
			if ((st.st_rdev & 0xFFC0) != 0x0400) continue;
#endif
			/* Root is always OK. */
			if (strcmp(ut->ut_user, "root") == 0) {
				user_ok++;
				break;
			}

			/* See if this is an allowed user. */
			for(i = 0; i < 32 && downusers[i]; i++)
				if (!strncmp(downusers[i], ut->ut_user,
				    UT_NAMESIZE)) {
					user_ok++;
					break;
				}
		}
		endutent();

		/* See if user was allowed. */
		if (!user_ok) {
			if ((fp = fopen(CONSOLE, "w")) != NULL) {
				fprintf(fp, "\rshutdown: no authorized users "
						"logged in.\r\n");
				fclose(fp);
			}
			exit(1);
		}
	}

	/* Read pid of running shutdown from a file */
	if ((fp = fopen(SDPID, "r")) != NULL) {
		if (fscanf(fp, "%d", &pid) != 1)
			pid = 0;
		fclose(fp);
	}

	/* Read remaining words, skip time if needed. */
	message[0] = 0;
	for(c = optind + (!cancel && !when[0]); c < argc; c++) {
		if (strlen(message) + strlen(argv[c]) + 4 > MESSAGELEN)
			break;
  		strcat(message, argv[c]);
  		strcat(message, " ");
	}
	if (message[0]) strcat(message, "\r\n");

	/* See if we want to run or cancel. */
	if (cancel) {
		if (pid <= 0) {
			fprintf(stderr, "shutdown: cannot find pid "
					"of running shutdown.\n");
			exit(1);
		}
		init_setenv("INIT_HALT", NULL);
		if (kill(pid, SIGINT) < 0) {
			fprintf(stderr, "shutdown: not running.\n");
			exit(1);
		}
		if (message[0]) wall(message, 0);
		exit(0);
	}
  
	/* Check syntax. */
	if (when[0] == '\0') {
		if (optind == argc) usage();
                strncpy(when, argv[optind++], WHEN_SIZE - 1);
	}

	/* See if we are already running. */
	if (pid > 0 && kill(pid, 0) == 0) {
		fprintf(stderr, "\rshutdown: already running.\r\n");
		exit(1);
	}

	/* Extra check. */
	if (doself && down_level[0] != '0' && down_level[0] != '6') {
		fprintf(stderr,
		"shutdown: can use \"-n\" for halt or reboot only.\r\n");
		exit(1);
	}

	/* Tell users what we're gonna do. */
	switch(down_level[0]) {
		case '0':
			strncpy(newstate, "for system halt", STATELEN);
			break;
		case '6':
			strncpy(newstate, "for reboot", STATELEN);
			break;
		case '1':
			strncpy(newstate, "to maintenance mode", STATELEN);
			break;
		default:
			snprintf(newstate, STATELEN, "to runlevel %s", down_level);
			break;
	}

	/* Go to the root directory */
	if (chdir("/")) {
		fprintf(stderr, "shutdown: chdir(/): %m\n");
		exit(1);
	}

	/* Create a new PID file. */
	unlink(SDPID);
	umask(022);
	if ((fp = fopen(SDPID, "w")) != NULL) {
		fprintf(fp, "%d\n", getpid());
		fclose(fp);
	} else if (errno != EROFS)
		fprintf(stderr, "shutdown: warning: cannot open %s\n", SDPID);

	/*
	 *	Catch some common signals.
	 */
	signal(SIGQUIT, SIG_IGN);
	signal(SIGCHLD, SIG_IGN);
	signal(SIGHUP,  SIG_IGN);
	signal(SIGTSTP, SIG_IGN);
	signal(SIGTTIN, SIG_IGN);
	signal(SIGTTOU, SIG_IGN);

	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = stopit;
	sigaction(SIGINT, &sa, NULL);

	if (fastboot)  close(open(FASTBOOT,  O_CREAT | O_RDWR, 0644));
	if (forcefsck) close(open(FORCEFSCK, O_CREAT | O_RDWR, 0644));

	/* Alias now and take care of old '+mins' notation. */
	if (!strcmp(when, "now")) strcpy(when, "0");

        sp = when;
	/* Validate time argument. */
	for ( ; *sp; sp++) {
		if (*sp != '+' && *sp != ':' && (*sp < '0' || *sp > '9'))
			usage();
	}
	sp = when;
	/* Decode shutdown time. */
	if (when[0] == '+') sp++;
	if (strchr(when, ':') == NULL) {
		/* Time in minutes. */
		wt = atoi(sp);
		if (wt == 0 && sp[0] != '0') usage();
	} else {
		if (sscanf(when, "%d:%2d", &hours, &mins) != 2) usage();
		/* Time in hh:mm format. */
		if (when[0] == '+') {
			/* Hours and minutes from now */
			if (hours > 99999 || mins > 59) usage();
			wt = (60*hours + mins);
			if (wt < 0) usage();
		} else {
			/* Time of day */
			if (hours > 23 || mins > 59) usage();
			time(&t);
			lt = localtime(&t);
			wt = (60*hours + mins) - (60*lt->tm_hour + lt->tm_min);
			if (wt < 0) wt += 1440;
		}
	}
	/* Shutdown NOW if time == 0 */
	if (wt == 0) issue_shutdown(halttype);

        /* Rather than loop and reduce wt (wait time) once per minute,
           we shall check the current time against the target time.
           Then calculate the remaining waiting time based on the difference
           between current time and target time.
           This avoids missing shutdown time (target time) after the
           computer has been asleep. -- Jesse
        */
        /* target time, in seconds = current time + wait time */
        time(&t);
        target_time = t + (60 * wt); 

	/* Give warnings on regular intervals and finally shutdown. */
	if (wt < 15 && !needwarning(wt, quiet_level)) issue_warn(wt);
	while(wt) {
		if (wt <= 5 && !didnolog) {
			donologin(wt);
			didnolog++;
		}
		if (needwarning(wt, quiet_level)) issue_warn(wt);
		hardsleep(60);
                time(&t);    /* get current time once per minute */
                if (t >= target_time)     /* past the target */
                  wt = 0;
                else if ( (target_time - t) <= 60 )  /* less 1 min remains */
                {
                    hardsleep(target_time - t);
                    wt = 0;
                }
                else                      /* more than 1 min remains */
                   wt = (int) (target_time - t) / 60;
	}
	issue_shutdown(halttype);

	return 0; /* Never happens */
}
